h/tasks/indexer.py

```diff
@@ -43,3 +47,23 @@ def delete_annotation(id_):
 def sync_annotations(limit):
     search_index = celery.request.find_service(name="search_index")
     search_index.sync(limit)
+
+
+@celery.task(acks_late=False)
+def report_sync_annotations_queue_length():
+    now = datetime.utcnow()
+
+    queue_length = (
+        celery.request.db.query(Job)
+        .filter(
+            Job.name == "sync_annotation",
+            Job.tag.in_(["storage.create_annotation", "storage.update_annotation"]),
+            Job.scheduled_at < now,
+            Job.expires_at > now,
+        )
+        .count()
+    )
+
+    newrelic.agent.record_custom_metric(
+        "Custom/SyncAnnotationsQueueLength", queue_length
+    )
```

Comment:
I considered moving this into the `Queue` class itself instead of doing it directly in the Celery task. It could even be `Queue.__len__`. But it's very specific in that it only counts `"sync_annotation"` tasks with the tags `"storage.create_annotation"` or `"storage.update_annotation"`.

I wonder if `Queue` could have a `len()` method like this?

```python
class Queue:
    ...

    def len(self, name=None, tags=None):
        now = datetime.utcnow()

        query = celery.request.db.query(Job).filter(
            Job.scheduled_at < now,
            Job.expires_at > now,
        )

        if name:
            query = query.filter(Job.name == "sync_annotation")

        if tags:
            query = query.filter(Job.tag.in_([tags]))

        return query.count()

    def __len__(self):
        return self.len()
```

Then the Celery task could call `queue.len(name="sync_annotation", tags=["storage.create_annotation", "storage.update_annotation"])`.

This would also leave room for counting other kinds of jobs in the future. For example I can imagine us wanting a count of _all_ scheduled and non-expired jobs, a count of not-yet-scheduled jobs, a count of expired jobs.

Reply:
At a minimum we need to:

* Filter in only `name="sync-annotations"` jobs. Those are the only type of jobs currently. But I specifically want to create an alarm for the sync-annotations job queue and don't want another type of job to trigger the alarm if added in the future
* Filter in only `tag="storage.create_annotation"` and `tag="storage.update_annotation"` jobs. I want the alarm to go off if the jobs added by the API aren't getting eaten fast enough for some reason. But I don't want a large batch of jobs added by an admin page to trigger the alarm
* Filter out expired jobs. These get ignored (https://github.com/hypothesis/h/pull/6371) so I don't want them to contribute to the alarm. (I'm probably going to add a separate metric for counting the expired jobs, and a separate alarm for them, in a later PR)
* Optional and not crucial: filter out jobs that aren't scheduled yet

Reply:
I've changed this to a `Queue.count(tags=None)` method and then the Celery task calls `queue.count(tags=["storage.create_annotation", "storage.update_annotation"])` and reports the result to New Relic. In the future I think we might want to report `queue.count()` (all tags) to New Relic, and I think we might also want to report the count of expired jobs and maybe also the count of not-yet-scheduled jobs (both of which would require some minor code tweaks to add more options to the `count()` method).